Add new ActionComponent helper/interfaces to facilitate short component input syntax

GiudGiud opened the pull request:
refs #29856

so I think PhysicsComponentInterface is actually a nice interface
but ComponentPhysics/IC/BC/MaterialPropertiesInterfaces are actually more helpers than interfaces? Maybe should be renamed to ComponentPhysics/.../...Helper ?

there are a few things we could imagine adding:
- making sure the surfaces for the BCs by the user actually belong to the component
if they are mispelled, they will hit an error later anyway in the boundary condition being added. But the earlier the better since context is lost by the action
- adding ICs on a per-component-block basis rather than on the entire component. This can be added later as the default of "define everywhere on the component" would be kept imo

GiudGiud committed:
Rename PhysicsComponentBase to ComponentPhysicsInterface

GiudGiud committed:
Add a ComponentMaterialPropertyInterface which lets us define material properties
for each component. This will let Physics query the properties to use
We may want to let Physics request the addition of a material that performs the conversion from functor to material property

Add this new interface to framework components which serve as demonstrations
refs #29856

GiudGiud committed:
Add tests for the ComponentMaterialPropertyInterface:
- exception for input size
- regression against a Physics-defined properties case
refs #29856

GiudGiud committed:
Add more info in the FEProblemBase material coverage check

GiudGiud committed:
Add an interface to define ICs directly on components
refs #29856

GiudGiud committed:
Add componentIC interface to the basic action components, for convenience/demonstration purposes
refs #29856

GiudGiud committed:
Add a test for defining ICs directly on (action)components
refs #29856

GiudGiud committed:
Simplify the ComponentPhysicsInterface

GiudGiud committed:
Add an interface to implement various default component behaviors on Physics
refs #29856

GiudGiud committed:
Add the PhysicsComponent interface to diffusion physics to enable adding ICs directly on components
refs #29856

GiudGiud committed:
Add a FunctorIC to be able to use the component-defined functors easier

GiudGiud committed:
Add the parameter check utils to action components

GiudGiud committed:
Add an interface/helper to set generic boundary conditions directly on components
refs #29856

GiudGiud committed:
Add support for adding boundary conditions from components in a Physics
refs #29856

GiudGiud committed:
Add new capability for adding BCs from component to cylinderComp and MGComp
refs #29856

GiudGiud committed:
Add new capability for adding BCs from component to the DIffusionCG Physics
refs #29856

GiudGiud committed:
Add a test for setting BCs directly on components
refs #29856

GiudGiud committed:
Add documentation for new interfaces
refs #29856

GiudGiud committed:
Address test failures:
- bad cast
- missing newly required parameters
- bad doco link

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2691677/), step Docs: sync website on 660f72d wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29880/site)

This comment will be updated on new commits.

GiudGiud committed:
Add list of defined blocks for functor material properties in error message

GiudGiud committed:
Add option for subdomains on ElementGenerator to facilitate 0D component mesh generation

GiudGiud committed:
Add a check integrity routine for late checks in PhysicsBase

GiudGiud committed:
Misc fixups for diffusionPhysics (handle AD for postprocessing)
Improve error messages

GiudGiud committed:
Add a routine to know if the problem has a solver variable, and make it accessible in Physics

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2698334/), step Fixup commit check on c76391a wanted to post the following:

The following commit(s) contain fixups:

- 100af03635: fixup! Add documentation for new interfaces refs #29856

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2698455/), step Generate coverage on 020a1c2 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b05d63</th>
      <th colspan=4>#29880 020a1c</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.29%</td>
      <td>85.27%</td>
      <td>-0.02%</td>
      <td>78.57%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108456</td>
      <td>108764</td>
      <td>+308</td>
      <td>330</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18711</td>
      <td>18790</td>
      <td>+79</td>
      <td>90</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29880/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29880/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29880/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 78.57% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud commented:
I'll improve coverage with a test Physics object but otherwise this is ready

GiudGiud committed:
Improve an assert and an error message on a missing parameter

lindsayad commented on framework/include/actioncomponents/ActionComponent.h:
```diff
@@ -78,12 +84,15 @@ class ActionComponent : public Action
+    return *_awh.problemBase().get();
   }
 
+  /// Get the factory to build (often physics-related but not always) objects (for example a Positions)
+  virtual Factory & getFactory() const { return _factory; }
```
There are derived objects that would override this?

GiudGiud replied:
will remove virtual. this was copied from PhysicsBase. I ll fix it there

lindsayad commented on framework/include/actioncomponents/ComponentBoundaryConditionInterface.h:
```diff
@@ -0,0 +1,98 @@
+  };
+
+  /**
+   * Whether the component has a boundary condition parameter is specified for the requested
```
```suggestion
   * Whether the component has a boundary condition parameter specified for the requested
```
Fix the language throughout this file

lindsayad commented on framework/include/actioncomponents/ComponentBoundaryConditionInterface.h:
```diff
@@ -0,0 +1,98 @@
+   * message
+   * @return the name of the boundaries
+   */
+  std::vector<BoundaryName> getBoundaryConditionsBoundaries(const VariableName & variable) const;
```
```suggestion
  std::vector<BoundaryName> getBoundaryConditionBoundaries(const VariableName & variable) const;
```

lindsayad commented on framework/include/actioncomponents/ComponentBoundaryConditionInterface.h:
```diff
@@ -0,0 +1,98 @@
+   * @param bc_type the type of the boundary condition (flux or fixed value)
+   * @return the name of the functor providing the value of the boundary condition
+   */
+  const MooseFunctorName getBoundaryCondition(const VariableName & variable,
```
Either return by non-const value or const reference

GiudGiud replied:
will fix

lindsayad commented on framework/include/actioncomponents/ComponentBoundaryConditionInterface.h:
```diff
@@ -0,0 +1,98 @@
+  const MooseFunctorName getBoundaryCondition(const VariableName & variable,
+                                              const BoundaryName & boundary,
+                                              const std::string & requestor_name,
+                                              BoundaryConditionType & bc_type) const;
```
Is this changed in the method?

GiudGiud replied:
yup this is effectively a return

want both returns to be argu,ments?

lindsayad replied:
It doesn't really matter to me, just as long as it's clear that `bc_type` is an "output" in the doxygen

lindsayad commented on framework/include/actioncomponents/ComponentBoundaryConditionInterface.h:
```diff
@@ -0,0 +1,98 @@
+  // This would let us return the boundary condition functor name by reference too
+
+  /// Maps of the fixed value boundary conditions
+  std::unordered_map<std::string, std::unordered_map<std::string, std::string>> _fixed_value_bcs;
+  /// Maps of the flux boundary conditions
+  std::unordered_map<std::string, std::unordered_map<std::string, std::string>> _flux_bcs;
```
Hashing strings is O(N). Unless the maps are going to be huge I'd change these to `std::map`

GiudGiud replied:
changed

lindsayad commented on framework/include/physics/DiffusionPhysicsBase.h:
```diff
@@ -23,7 +24,7 @@ class ActionComponent;
  * equation
  */
-class DiffusionPhysicsBase : public PhysicsBase
+class DiffusionPhysicsBase : virtual public PhysicsBase, virtual public PhysicsComponentInterface
```
Virtual inheritance of `PhysicsBase` too now? Why?

lindsayad replied:
I see with the inheritance of `PhysicsBase` in `PhysicsComponentInterface`

GiudGiud replied:
what do you think ? is it worth the additional complexity of virtual inheritance?

lindsayad replied:
You're the one developing most in these systems so I think you're the most qualified to make that decision

GiudGiud replied:
ok I ll keep with virtual inheritance. 
The alternative is templating and I think that s a tad harder to work with though less confusing

lindsayad commented on framework/include/physics/PhysicsComponentInterface.h:
```diff
@@ -0,0 +1,77 @@
+protected:
+  // TODO: add hash constructor to ComponentName to be able to use it as a key
+  /// Map of components to variables and initial conditions
+  std::unordered_map<std::string, std::map<VariableName, MooseFunctorName>>
+      _components_initial_conditions;
+  /// Map of components to variables and boundary conditions
+  std::unordered_map<
```
same concern regarding hashing strings

GiudGiud replied:
fixed

lindsayad commented on framework/include/utils/PiecewiseByBlockLambdaFunctor.h:
```diff
@@ -162,6 +166,11 @@ PiecewiseByBlockLambdaFunctor<T>::setFunctor(const MooseMesh & libmesh_dbg_var(m
+
+  // Handle special case of ANY_BLOCK_ID and empty block restriction that also cover
+  // INVALID_BLOCK_ID
+  if (block_ids.count(Moose::ANY_BLOCK_ID) || block_ids.empty())
+    add_lammy(Moose::INVALID_BLOCK_ID);
```
For `ANY_BLOCK_ID` in particular, should we loop over the mesh subdomains and add the lambda for every block?

lindsayad commented on framework/include/utils/PiecewiseByBlockLambdaFunctor.h:
```diff
@@ -198,15 +207,25 @@ PiecewiseByBlockLambdaFunctor<T>::hasBlocks(const SubdomainID id) const
 {
+  std::vector<SubdomainID> block_ids;
+  block_ids.reserve(functors.size());
+  for (const auto & [key, value] : functors)
+  {
+    libmesh_ignore(value);
+    block_ids.push_back(key);
```
```suggestion
  for (const auto & [available_sub_id, functor] : functors)
  {
    libmesh_ignore(functor);
    block_ids.push_back(available_sub_id);
```

lindsayad commented on framework/src/actioncomponents/ComponentPhysicsInterface.C:
```diff
@@ -7,29 +7,29 @@
   : ActionComponent(params), _physics_names(getParam<std::vector<PhysicsName>>("physics"))
 {
-  // Should be done later?
+  // Adding ActionComponents must be done after adding adding Physics
```
```suggestion
  // Adding ActionComponents must be done after adding Physics
```

lindsayad commented on framework/src/physics/DiffusionCG.C:
```diff
@@ -61,10 +65,14 @@ DiffusionCG::addFEKernels()
       params.set<MaterialPropertyName>("diffusivity") =
           getParam<MaterialPropertyName>("diffusivity_matprop");
     else if (isParamValid("diffusivity_functor"))
+      params.set<FunctionName>("function") = getParam<MooseFunctorName>("diffusivity_functor");
```
Functors aren't always functions. How does this work?

GiudGiud replied:
see above, there is an error if the name is not the name of a function in the problem

lindsayad commented on framework/src/physics/DiffusionPhysicsBase.C:
```diff
@@ -124,15 +127,53 @@ void
+{
+  InputParameters params = getFactory().getValidParams("FunctorIC");
+
+  for (const auto & comp_pair : _components_initial_conditions)
```
You use both members of the pair, so let's use a structured binding for better readability

GiudGiud replied:
done

lindsayad commented on framework/src/variables/MooseVariableFE.C:
```diff
@@ -863,7 +863,8 @@ typename MooseVariableFE<OutputType>::ValueType
 {
   mooseAssert(node_arg.node, "Must have a node");
-  mooseAssert(this->hasBlocks(node_arg.subdomain_id),
+  mooseAssert(node_arg.subdomain_id == Moose::INVALID_BLOCK_ID ||
```
You must be using `Moose::INVALID_BLOCK_ID` because of something I've done? The assert definitely looked better before. Changing this to `Moose::ANY_BLOCK_ID` would seem more correct, but again I'm guessing you've added this code because of something preexisting that I wrote. I wonder if we can do better

GiudGiud replied:
in nodal kernels we tend not to know what subdomain id to use unfortunately. so I ve started writing nodal args with invalid block id, but then they cant be used for this

lindsayad replied:
If that's the case, I don't think functors should accept `INVALID_BLOCK_ID` as valid input. If an object using a functor isn't picky, I think putting `ANY_BLOCK_ID` as the subdomain ID is better

GiudGiud replied:
interesting. That works for me!

lindsayad commented on test/tests/actioncomponents/interfaces/component_with_bcs.i:
```diff
@@ -0,0 +1,170 @@
+    dy = 1
+    subdomain_id = 2
+  []
+  final_generator = 'cmg'
```
Is this line necessary? Something to do with the meshes coming from components?

lindsayad replied:
hopefully that's something we can address in the future

lindsayad commented on test/tests/actioncomponents/interfaces/component_with_bcs.i:
```diff
@@ -0,0 +1,170 @@
+    direction = '0 1 0'
+    block = 'cyl1'
+
+    physics = 'diff_u'
```
This is really nice. I like it!

lindsayad commented on test/tests/actioncomponents/interfaces/tests:
```diff
@@ -0,0 +1,73 @@
+    csvdiff = 'component_with_physics_out.csv'
+    design = 'ComponentPhysicsInterface.md'
+    issues = '#24103'
+    requirement = 'The system shall be able to add equations onto the mesh of (action)components using the Physics syntax.'
```
```suggestion
    requirement = 'The system shall be able to add equations onto the mesh of (action)components using physics syntax.'
```

lindsayad reviewed (commented):
Overall looks quite good to me

joshuahansel commented on framework/src/actioncomponents/CylinderComponent.C:
```diff
@@ -12,14 +12,22 @@
 #include "RotationMatrix.h"
 
 registerMooseAction("MooseApp", CylinderComponent, "add_mesh_generator");
+// CylinderComponent is an exmaple of ComponentPhysicsInterface
```
```suggestion
// CylinderComponent is an example of ComponentPhysicsInterface
```

GiudGiud committed:
Address Alex's review

- better names for routines
- fixup docstrings
- use map for string over unstructured map
- use structured binding when looping over component BC maps

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

joshuahansel commented on framework/src/ics/FunctorIC.C:
```diff
@@ -0,0 +1,77 @@
+    _scaling(getParam<Real>("scaling_factor")),
+    _pl(_fe_problem.mesh().getPointLocator())
+{
+  // Check supported an unsupported functors
```
```suggestion
  // Check supported and unsupported functors
```

joshuahansel commented on framework/src/ics/FunctorIC.C:
```diff
@@ -0,0 +1,77 @@
+{
+  // Check supported an unsupported functors
+  const auto & functor_name = getParam<MooseFunctorName>("functor");
+  // PPs: need to execute before
```
```suggestion
  // See https://github.com/idaholab/moose/issues/19396 for discussion of the functor restrictions.
  // PPs: need to execute before
```

joshuahansel commented on framework/doc/content/source/actioncomponents/ComponentBoundaryConditionInterface.md:
```diff
@@ -0,0 +1,17 @@
+- the [!param](/ActionComponents/CylinderComponent/fixed_value_bc_variables), [!param](/ActionComponents/CylinderComponent/fixed_value_bc_boundaries) and [!param](/ActionComponents/CylinderComponent/fixed_value_bc_values) to define fixed value / Dirichlet boundary conditions on the surfaces of the `ActionComponent`
+- the [!param](/ActionComponents/CylinderComponent/flux_bc_variables), [!param](/ActionComponents/CylinderComponent/flux_bc_boundaries) and [!param](/ActionComponents/CylinderComponent/flux_bc_values) to define flux boundary conditions on the surfaces of the `ActionComponent`
+
+An [ActionComponent.md] inheriting `ComponentBoundaryConditionInterface` must be registered to the `init_component_physics` and `check_integrity`
```
Optional: do we have any way to check/enforce this?

GiudGiud replied:
we do in Physics with the addRequiredTask routine

GiudGiud replied:
and in ActionComponent
```
  /// Add a new required task for all physics deriving from this class
  /// NOTE: This does not register the task, you still need to call registerMooseAction
  void addRequiredTask(const std::string & task) { _required_tasks.insert(task); }
```

I ll make sure it's used

GiudGiud replied:
done

joshuahansel reviewed (changes requested):
Looks good! Just address the few comments

GiudGiud committed:
Address Josh's review

- request the needed tasks in interfaces
- fix some typos
- add a link for more info on functor types

Co-authored-by: joshuahansel <joshua.hansel@inl.gov>

GiudGiud committed:
Add routines for retrieving property functors from components with the interface

GiudGiud committed:
Add testing for the functorIC
refs #29980

GiudGiud committed:
Add more coverage:
- add a test for diffusive flux output
- use diffusivity functor in CG in a test
- add a subdomain name to element generator (0, same as subdomain id)
- add a test physics to check on the unimplemented physics-component interactions error messages

GiudGiud commented:
comments should be addressed now and coverage should be better

GiudGiud commented:
@lindsayad please let me know if you want to take another look

GiudGiud committed:
Add block restrictions to kernels and ics for heat conduction -cg
Add a nonAD mode for the heat conduction CG Physics: for a case that wont converge with AD
Add a test for no-AD mode

GiudGiud commented:
@joshuahansel added a quick commit on top to help with TMAP8 work

GiudGiud commented:
> Do you know why it's not converging with AD?

not really. It's only in parallel, and with MUMPs-LU

lindsayad commented:
Running your test with `use_automatic_differentiation=true` is working fine for me for 2,4,9 process counts with mumps on Linux. Are you running the tests on Mac?

GiudGiud commented:
the bug is in the TMAP 8 input not here

GiudGiud commented:
cleaning up the TMAp8 PR then posting a link

GiudGiud commented:
TMAP8 input is there. You need to build MOOSE with this branch to run it
https://github.com/idaholab/TMAP8/pull/139/files#diff-8d1f0a29b9211b9dc77825bff0c2681592d43f69385f5f384c7de6f62344c658

lindsayad commented:
Ok I was able to reproduce the `FACTOR_OUTMEMORY` with MUMPS. You can fix this by passing `-mat_mumps_icntl_14 300` which instructs MUMPS to increase the factorization working space memory by 300%

GiudGiud commented:
Only changes since review are rebase, and deleting a commit thanks to new options from #29971
